File: types/anime.ts
export interface KodikTranslation {
  title: string,
}

export interface KodikMaterialData {
  description: string,
  poster_url: string,
  anime_poster_url: string,
  screenshots: string[],
  episodes_total: number | null,
  episodes_aired: number | null,
}

export interface KodikInfo {
  id: string,
  title: string,
  link: string,
  kinopoisk_id: string | null,
  translation: KodikTranslation,
  screenshots: string[],
  material_data?: KodikMaterialData,
}

export interface ShikimoriInfo {
  id: number;
  name: string;
  russian: string;
  poster: {
    mainUrl: string;
  };
  kind: string;
  score: number;
  airedOn: {
    date: string;
  };
  franchise: string;
}

export interface Genre {
  id: number;
  name: string;
  russian: string;
  kind: string;
}

export interface Studio {
  id: number;
  name: string;
  filtered_name: string;
  real: boolean;
  image: string;
}

export interface Video {
  id: number;
  url: string;
  image_url: string;
  player_url: string;
  name: string;
  kind: string;
  hosting: string;
}

export interface Screenshot {
  original: string;
  preview: string;
}

export interface WatchHistoryItem {
  animeId: number;
  title: string;
  image: string;
  link?: string;
  lastWatched: number;
}

export const MISSING_POSTER_URL = 'https://shikimori.one/assets/globals/missing_preview.jpg';

export const KIND_PRIORITY: Record<string, number> = {
  'tv': 0,
  'movie': 1,
  'ova': 2,
  'ona': 3,
  'special': 4,
  'tv_special': 5,
  'web': 6,
  'pv': 7,
};

const HIDE_KINDS = ['music', 'cm'];
const SINGLE_KINDS = ['movie', 'special']

export function canShow(anime: ShikimoriInfo) {
  return !HIDE_KINDS.includes(anime.kind) && anime.kind !== null;
}

export function canShowSeries(anime: ShikimoriInfo) {
  return !SINGLE_KINDS.includes(anime.kind);
}

export function canOpen(anime: ShikimoriInfo) {
  return anime.score !== 0;
}

export type AnimeStatus = 'watching' | 'planned' | 'completed' | 'on_hold' | 'dropped';

export const STATUS_COLORS: Record<AnimeStatus, string> = {
  watching: '#4CAF50', 
  planned: '#2196F3', 
  completed: '#9C27B0', 
  on_hold: '#FF9800', 
  dropped: '#F44336', 
};

export const STATUS_LABELS: Record<AnimeStatus, string> = {
  watching: 'Смотрю',
  planned: 'В планах',
  completed: 'Просмотрено',
  on_hold: 'Отложено',
  dropped: 'Брошено',
};

File: lib/models/anime.dart
/// Порт `types/anime.ts`. Значения констант и логика — без изменений.
library;

/// Заглушка постера, когда у аниме нет своей картинки.
const String missingPosterUrl =
    'https://shikimori.io/assets/globals/missing_preview.jpg';

/// Приоритет вида (kind) при сортировке связанных тайтлов.
const Map<String, int> kindPriority = {
  'tv': 0,
  'movie': 1,
  'ova': 2,
  'ona': 3,
  'special': 4,
  'tv_special': 5,
  'web': 6,
  'pv': 7,
};

const List<String> _hideKinds = ['music', 'cm'];
const List<String> _singleKinds = ['movie', 'special'];

/// Модель тайтла Shikimori.
///
/// JSON от Shikimori содержит вложенные объекты `poster: { mainUrl }`
/// и `airedOn: { date }`, каждый из которых может быть `null`.
class ShikimoriInfo {
  const ShikimoriInfo({
    required this.id,
    required this.name,
    this.russian,
    this.posterUrl,
    this.kind,
    this.score = 0,
    this.airedOnDate,
  });

  final int id;
  final String name;
  final String? russian;
  final String? posterUrl;
  final String? kind;
  final double score;
  final String? airedOnDate;

  /// Принимает любой `Map`, а не только `Map<String, dynamic>`.
  ///
  /// Hive возвращает вложенные объекты как `Map<dynamic, dynamic>`, а
  /// `Map<String, dynamic>.from()` копирует лишь верхний уровень. Жёсткий каст
  /// вложенного `poster` из-за этого падал при чтении кэша — приложение
  /// открывалось пустым экраном до очистки данных.
  factory ShikimoriInfo.fromJson(Map<dynamic, dynamic> json) {
    final poster = json['poster'];
    final airedOn = json['airedOn'];
    final rawScore = json['score'];
    return ShikimoriInfo(
      id: json['id'] is int ? json['id'] as int : int.tryParse('${json['id']}') ?? 0,
      name: json['name']?.toString() ?? '',
      russian: json['russian']?.toString(),
      posterUrl: poster is Map ? poster['mainUrl']?.toString() : null,
      kind: json['kind']?.toString(),
      score: rawScore == null
          ? 0
          : (rawScore is num ? rawScore.toDouble() : double.tryParse('$rawScore') ?? 0),
      airedOnDate: airedOn is Map ? airedOn['date']?.toString() : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'russian': russian,
        'poster': {'mainUrl': posterUrl},
        'kind': kind,
        'score': score,
        'airedOn': {'date': airedOnDate},
      };

  ShikimoriInfo copyWith({
    int? id,
    String? name,
    String? russian,
    String? posterUrl,
    String? kind,
    double? score,
    String? airedOnDate,
  }) {
    return ShikimoriInfo(
      id: id ?? this.id,
      name: name ?? this.name,
      russian: russian ?? this.russian,
      posterUrl: posterUrl ?? this.posterUrl,
      kind: kind ?? this.kind,
      score: score ?? this.score,
      airedOnDate: airedOnDate ?? this.airedOnDate,
    );
  }
}

bool canShow(ShikimoriInfo anime) =>
    anime.kind != null && !_hideKinds.contains(anime.kind);

bool canShowSeries(ShikimoriInfo anime) => !_singleKinds.contains(anime.kind);

bool canOpen(ShikimoriInfo anime) => anime.score != 0;
